### PR TITLE
GH-73435: Implement recursive wildcards in `pathlib.PurePath.match()`

### PR DESCRIPTION
**Pull request by barneygale**
`PurePath.match()` now handles the `**` wildcard as in `Path.glob()`, i.e. it matches any number of path segments.

We now compile a `re.Pattern` object for the entire pattern. This is made more difficult by `fnmatch` not treating directory separators as special when evaluating wildcards (`*`, `?`, etc), and so we arrange the path parts onto separate *lines* in a string, and ensure we don't set `re.DOTALL`.

This improves performance of `match()` around 2x-3x times for simple patterns, and more for complex patterns:

```shell
$ ./python -m timeit \
    -s 'from pathlib import PureWindowsPath as P; path = P("C:/foo/bar.py"); pattern = P("c:/*/*.py")' \
    'path.match(pattern)'
50000 loops, best of 5: 8.13 usec per loop   # before
1000000 loops, best of 5: 297 nsec per loop  # after
```

* Issue: gh-73435

### CODE REVIEW

**Comment by barneygale** [comment]
Two big caveats:

- This doesn't support paths or patterns with embedded newlines
- #101360 must be fixed first

File: Lib/pathlib.py [hunk 1]
```diff
@@ -61,6 +61,15 @@ def _is_case_sensitive(flavour):
 # Globbing helpers
 #
 
+
+_FNMATCH_PREFIX, _FNMATCH_SUFFIX = fnmatch.translate('_').split('_')
+_FNMATCH_SLICE = slice(len(_FNMATCH_PREFIX), -len(_FNMATCH_SUFFIX))
```
Comment (barneygale) [inline Lib/pathlib.py:74]: This is a bit of a hack. Alternatively I could add an argument to `fnmatch.translate()` that makes it exclude the prefix/suffix, or add a new private `fnmatch._translate()` function perhaps? Not sure.

**Comment by hugovk** [comment]
Let's also add it to the What's New.

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -365,6 +365,9 @@ pathlib
 * Add :meth:`pathlib.Path.is_junction` as a proxy to :func:`os.path.isjunction`.
   (Contributed by Charles Machalow in :gh:`99547`.)
 
+* Add support for recursive wildcards in :meth:`pathlib.PurePath.match`.
+  (Contributed by Barney Gale in :gh:`101398`.)
```
Comment (hugovk) [inline Doc/whatsnew/3.12.rst]: Let's use the issue number:
```suggestion
  (Contributed by Barney Gale in :gh:`73435`.)
```

**Comment by barneygale** [comment]
Thank you for the feedback, Hugo!

File: Lib/pathlib.py (hunk 1 above)
Comment (barneygale) [reply Lib/pathlib.py:74]: @AlexWaygood any recommendations? Thanks

File: Lib/pathlib.py [hunk 3]
```diff
@@ -430,6 +446,41 @@ def _parts_normcase(self):
...
+            self._lines_cached = str(self).translate(trans)
+            return self._lines_cached
+
+    @property
+    def _matcher(self):
+        try:
+            return self._matcher_cached
+        except AttributeError:
+            if not self.parts:
+                raise ValueError("empty pattern")
```
Comment (AlexWaygood) [inline Lib/pathlib.py]: This should use `raise from None`, to suppress exception chaining, I think -- otherwise the traceback might imply to a user that that there's a bug in `pathlib`

```suggestion
                raise ValueError("empty pattern") from None
```

File: Lib/pathlib.py [hunk 4]
```diff
@@ -430,6 +446,41 @@ def _parts_normcase(self):
...
+            parts = [r'\A' if self.drive or self.root else '^']
+            for part in self._lines.splitlines(keepends=True):
+                if part == '**\n':
+                    part = r'[\s\S]*^'
+                elif part == '**':
+                    part = r'[\s\S]*'
+                elif '**' in part:
+                    raise ValueError("Invalid pattern: '**' can only be an entire path component")
```
Comment (AlexWaygood) [inline Lib/pathlib.py]:
```suggestion
                    raise ValueError("Invalid pattern: '**' can only be an entire path component") from None
```

**Review by AlexWaygood — COMMENTED** [review]
Not what you asked about, but something I noticed while reading the patch -- I think some of the exception handling could be slightly improved; see my comments below

File: Lib/pathlib.py (hunk 3 above)
Comment (barneygale) [reply Lib/pathlib.py]: Good spot!

File: Lib/pathlib.py (hunk 1 above)
Comment (AlexWaygood) [reply Lib/pathlib.py:74]: Hrmmmmmmm. I don't mind a hack for an implementation detail, but this isn't particularly readable as-is :/

what about defining a new function here:

```suggestion
_FNMATCH_PREFIX, _FNMATCH_SUFFIX = fnmatch.translate('_').split('_')


def _fnmatch_translate_stripping_prefix(pat):
    """fnmatch.translate(), but with the prefix and suffix stripped from the regex
    
    >>> fnmatch.translate("_")
    '(?s:_)\\Z'
    >>> _fnmatch_translate_stripping_prefix("_")
    '_'
    """
    return fnmatch.translate(pat)[len(_FNMATCH_PREFIX) : -len(_FNMATCH_SUFFIX)]


```

And then you could use this new function on this line, where you currently do `fnmatch.translate(part)[_FNMATCH_SLICE]`. (The `len()` calls should take up almost no time at all, by the way, so I don't think there's any need for the `_FNMATCH_SLICE` variable, which makes it less readable imo.)

Comment (AlexWaygood) [reply Lib/pathlib.py:74]: I don't mind the current code _too_ much though. It took me a few seconds to grok what was going on, but I think it's perfectly fine to go ahead with if you don't like my suggestion :)

Comment (barneygale) [reply Lib/pathlib.py:74]: Makes sense. Do you think there's any merit in moving that function to `fnmatch._translate_unwrapped()` or similar? `fnmatch.translate()` could be implemented on top, and fnmatch wouldn't add the prefix/suffix only for pathlib to strip it again. OTOH, pathlib would use a private function from another module, so probably not ideal?

Comment (barneygale) [reply Lib/pathlib.py:74]: i.e.:

```diff
diff --git a/Lib/fnmatch.py b/Lib/fnmatch.py
index d5e296f774..28fd843eea 100644
--- a/Lib/fnmatch.py
+++ b/Lib/fnmatch.py
@@ -76,7 +76,11 @@ def translate(pat):
 
     There is no way to quote meta-characters.
     """
+    res = _translate_unwrapped(pat)
+    return fr'(?s:{res})\Z'
+
 
+def _translate_unwrapped(pat):
     STAR = object()
     res = []
     add = res.append
@@ -181,5 +185,4 @@ def translate(pat):
         else:
             add(f"(?>.*?{fixed})")
     assert i == n
-    res = "".join(res)
-    return fr'(?s:{res})\Z'
+    return "".join(res)
```

Comment (AlexWaygood) [reply Lib/pathlib.py:74]: I think it would be best not to use an private function of `fnmatch` from `pathlib`, if possible. That feels like it would create an unfortunate level of coupling between the two modules.

Instinctively I like your other idea of adding an `exclude_prefix` argument to `fnmatch.translate()`; it feels like a clean solution to our problem here. But we shouldn't bloat the `fnmatch` API unnecessarily just to fix our problem here, unless that extra argument is something that end users have actually asked for.

Comment (barneygale) [reply Lib/pathlib.py:74]: I suspect what I want here is fairly niche (unfortunately for me!).

I hesitate to add a `_fnmatch_translate_stripping_prefix` function to pathlib only because it will affect performance slightly IIUC. Perhaps I could add a comment above the `_FNMATCH_SLICE` gubbins to explain what's going on?

Comment (AlexWaygood) [reply Lib/pathlib.py:74]: I wonder if this hack might be more performant than a pass-through function written in Python:

```py
_strip_fnmatch_prefix_and_suffix = operator.itemgetter(slice(len(_FNMATCH_PREFIX), -len(_FNMATCH_SUFFIX)))
```

The `operator` functions are written in C, so that might be pretty fast.

And then lower down, where you currently do `fnmatch.translate(part)[_FNMATCH_SLICE]`, you could do `_strip_fnmatch_prefix_and_suffix(fnmatch.translate(part))` instead.

You'd have to import `operator` at the top of `pathlib`, but I think it may be imported at startup anyway, so it probably wouldn't slow down the import of `pathlib` too much.

Anyway, always worth doing some quick benchmarking! Adding a pass-through Python function here might not actually slow down the code too much; and using `operator` might not speed things up here.

Comment (barneygale) [reply Lib/pathlib.py:74]: Interestingly, using `slice()` without `itemgetter()` seems most performant:

```
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); prefix, suffix = translate("_").split("_")' 'pat[len(prefix):-len(suffix)]'
5000000 loops, best of 5: 86.3 nsec per loop
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); prefix, suffix = translate("_").split("_"); sl = slice(len(prefix), -len(suffix));' 'pat[sl]'
5000000 loops, best of 5: 54.1 nsec per loop
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); from operator import itemgetter; prefix, suffix = translate("_").split("_"); sl = itemgetter(slice(len(prefix), -len(suffix)))' 'sl(pat)'
5000000 loops, best of 5: 62.6 nsec per loop
```

Comment (barneygale) [reply Lib/pathlib.py:74]: I've added a comment giving an explanation for `_FNMATCH_SLICE` - do you think it's sufficient? I think the use of the `slice` object hints towards what's going on too.

Comment (AlexWaygood) [reply Lib/pathlib.py:74]: > I've added a comment giving an explanation for `_FNMATCH_SLICE` - do you think it's sufficient? I think the use of the `slice` object hints towards what's going on too.

Yes, I think that works fine! Won't formally approve as I haven't looked over the PR comprehensively (and don't have the time right now, I have language-summit blogs to finish this evening!!). But the comment looks great, and is very helpful :)

Comment (barneygale) [reply Lib/pathlib.py:74]: Thank you!

File: Lib/pathlib.py [hunk 5]
```diff
@@ -54,13 +54,29 @@ def _ignore_error(exception):
...
 # Globbing helpers
 #
 
+
+# fnmatch.translate() returns a regular expression that includes a prefix and
+# a suffix, which enable matching newlines and ensure the end of the string is
+# matched, respectively. These features are undesirable for our implementation
+# of PurePatch.match(), which represents path separators as newlines and joins
+# pattern segments together. As a workaround, we define a slice object that
+# can remove the prefix and suffix from any translate() result.
```
Comment (zooba) [inline Lib/pathlib.py]: I see the long discussion about this and think you've landed on something acceptable.

My preference would be for the `def _fnmatch_translate_stripping_prefix(pat)` in this file that Alex proposed (maybe `..._without_anchors`), as it expresses the intent more clearly at the point where it is used and keeps the details elsewhere.

But it's only a single use, and the slice is easy enough to find, so I wouldn't worry. Mostly I wanted to think out loud about architecture 😆

File: Lib/pathlib.py [hunk 6]
```diff
@@ -92,6 +108,26 @@ def _compile_pattern(pat, case_sensitive):
     return re.compile(fnmatch.translate(pat), flags).match
 
 
+@functools.lru_cache()
+def _compile_pattern_lines(pattern_lines, case_sensitive):
+    parts = ['^']
+    for part in pattern_lines.splitlines(keepends=True):
+        if part == '**\n':
+            part = r'[\s\S]*^'
```
Comment (zooba) [inline Lib/pathlib.py:134]: I assume the `[\s\S]` pattern implies something slightly more nuanced than `.`? Supposedly the two classes are the opposite of each other, and so should include everything.

A brief comment here will save the question being asked every time. (You can't over-explain your regex!)

File: Lib/pathlib.py (hunk 5 above)
Comment (zooba) [reply Lib/pathlib.py]: However, since this is the first comment being encountered, it would be nice to have a bit of an overview of how the new algorithm works. Just the kind of thing you've posted in various issues/PRs, and will likely be in the commit message, but better to have it near the code. I _would_ worry about this one.

**Review by zooba — COMMENTED** [review]
Approved, but consider adding a couple of comments (as suggested) so that the next person who has to trace through this code is grateful rather than mad at you ;-)

File: Lib/pathlib.py (hunk 5 above)
Comment (barneygale) [reply Lib/pathlib.py]: I've added a "see _compile_pattern_lines()" comment, and then expanded the explanation there. If you wanted, I could move that method directly below the slice definition?

File: Lib/pathlib.py (hunk 6 above)
Comment (barneygale) [reply Lib/pathlib.py:134]: Added a chonky comment! The important bit is that we're **not** running in `re.DOTALL` mode, and so `.` does not match newlines, which in our case represent path separators. It's this trick that allows `*` wildcards to not match path separators.

**Comment by zooba** [comment]
Perfect! Ship it

File: Lib/pathlib.py [hunk 7]
```diff
@@ -54,13 +54,30 @@ def _ignore_error(exception):
             getattr(exception, 'winerror', None) in _IGNORED_WINERRORS)
 
 
+@functools.lru_cache()
```
Comment (AlexWaygood) [inline Lib/pathlib.py]: Drive-by nit: since we know there's only two flavours, you can use an unbounded cache here. It should be a teeny tiny bit faster (since the cache mechanism doesn't have to worry about popping the least-recently-used cache entries out of the cache), and it's slightly prettier to look at as well :)

```suggestion
@functools.cache
```

**Comment by barneygale** [comment]
Thank you for your help Alex, Hugo and Steve!

**Comment by barneygale** [comment]
Hey, if it interests anyone, I have a follow-up PR that simplifies a bunch of the code added in this PR. It does this by adding a new *seps* parameter to `fnmatch.translate()`, and varying the generated regular expresison when its supplied.

- #106703
